napari_pyclesperanto_assistant/_gui/_Assistant.py

```diff
@@ -1,163 +1,118 @@
-
-        btn.setLayout(QGridLayout(btn))
-
-        icon_label = QLabel(btn)
-        icon_label.setAlignment(Qt.AlignCenter)
-        pixmap = QPixmap()
-        pixmap.load(str(Path(__file__).parent) + "/icons/" + title.lower().replace(" ", "_").replace("(", "").replace(")", "") + ".png")
-        pixmap = pixmap.scaled(QSize(40, 40), Qt.KeepAspectRatio)
-        icon_label.setPixmap(pixmap)
-        btn.layout().addWidget(icon_label)
-
-        text_label = QLabel(title, btn)
-        text_label.setAlignment(Qt.AlignCenter)
-        text_label.setWordWrap(True)
-        text_label.setFont(self.font)
-        btn.layout().addWidget(text_label)
-
-        def trigger():
-            self._activate(handler)
-
-        # action
-        btn.clicked.connect(trigger)
-        if x is None or y is None:
-            self.layout.addWidget(btn)
-        else:
-            self.layout.addWidget(btn, x, y)
-
-    def _activate(self, magicgui):
-        LayerDialog(self.viewer, magicgui)
-
-    def _export_jython_code(self):
-        generator = JythonGenerator(self.viewer.layers)
-        code = generator.generate()
-        self._save_code(code, default_fileending=generator.file_ending())
-
-    def _export_jython_code_to_clipboard(self):
-        generator = JythonGenerator(self.viewer.layers)
-        code = generator.generate()
-        import pyperclip
-        pyperclip.copy(code)
-
-    def _export_notebook(self, filename=None):
-        generator = PythonJupyterNotebookGenerator(self.viewer.layers)
-        code = generator.generate()
-        if filename is None:
-            filename = self._save_code(code, default_fileending=generator.file_ending())
-        if filename is not None:
-            import os
-            os.system('jupyter nbconvert --to notebook --inplace --execute ' + filename)
-            # os.system('jupyter notebook ' + filename) # todo: this line freezes napari
-
-    def _save_code(self, code, default_fileending = "*.*", filename = None):
-        if filename is None:
-            filename = QFileDialog.getSaveFileName(self, 'Save code as...', '.', default_fileending)
-        if filename[0] == '':
-            return None
-
-        filename = filename[0]
-        if not filename.endswith(default_fileending):
-            filename = filename + default_fileending
-
-        file = open(filename, "w+")
-        file.write(code)
-        file.close()
-
-        return filename
\ No newline at end of file
+        napari_viewer.layers.events.removed.connect(self._on_layer_removed)
+        napari_viewer.events.active_layer.connect(self._on_active_layer_change)
+        self._layers: Dict[Layer, Tuple[QtViewerDockWidget, FunctionGui]] = {}
+
+        self._grid = ButtonGrid(self)
+        self.setLayout(QVBoxLayout())
+        self.layout().addWidget(self._grid)
+        self._grid.addItems(CATEGORIES)
+        self._grid.itemClicked.connect(self._on_item_clicked)
+
+        # # create menu
```

Comment:
The code generators are currently not accessible, right? Shall we add a button to the button grid which opens this popup-menu? With that, we wouldn't pollute the main menu...

Reply:
yeah, not done yet... I've been thinking maybe rather than `Pipeline`, we should just use the `dask` format described here: https://docs.dask.org/en/latest/spec.html.
It's a pure-python representation of a task graph (that itself doesn't even require dask)... and then the nice thing is you can use [`dask.order.order`](https://docs.dask.org/en/latest/order.html) to make sure you execute the operations in the right order.  I just pushed a commit that adds a [`to_dask`](https://github.com/clEsperanto/napari_pyclesperanto_assistant/pull/10/files#diff-9089bb319a416b82bf455124b63948d6892ffb0779ccc63140f6942a9701365bR95) method to the Assistant, and a [`from_dask`](https://github.com/clEsperanto/napari_pyclesperanto_assistant/pull/10/files#diff-5c3db7bfe8c667dd4b51640f1872876e4f3793d9a203e7d35875756a8e0e7e35R83) method to the pipeline which can generate code from a dask task graph.

but regarding the menu, i don't think it's a problem to add one item to the Plugins menu (but it should be a single item with subactions rather than a bunch of top level actions)